python/phase-1.md

```diff
@@ -0,0 +1,380 @@
+
+## Overview
+
+The Python SDK will be backed by [sdk-core](https://github.com/temporalio/sdk-core), utilizing a bridge based on
+[PyO3](https://github.com/PyO3/PyO3) similar to `sdk-typescript`'s `neon`-based bridge.
+
+The high-level goals for the Python SDK are:
+
+* Be as simple as possible from a caller POV
+* Be as lightweight as possible wrt dependencies
+* Be reasonably flexible with reasonable defaults
+* Support only Python >= 3.7
+* Be friendly to REPL/interpreter use cases
+
+The stack/approach looks like:
+
+* Python >= 3.7 (so may need `typing_extension` for things like `ParamSpec`)
+* Poetry packaging, black formatting
+  * Pipenv originally considered
+* PyO3 + sdk-core
+* asyncio direct (i.e. no need for anyio-style wrapper)
+* Type annotations everywhere
+* Decorator-based activity and workflow definitions
+
+An example of a client (intentionally missing _a lot_ of features):
+
+```python
+import temporalio
+
+def main():
+  client = temporalio.Client("localhost:7233", namespace="MyNamespace")
+  result = await client.execute_workflow("MyWorkflow", "arg1", id="someid", task_queue="somequeue")
+  print("Result", result)
+```
+
+An example of an activity and registration (also missing a lot of features):
+
+```python
+import temporalio
+
+def say_hello(name: str) -> str:
+  return f"Hello, {name}!"
+
+def main(client: temporalio.Client):
+  worker = temporalio.Worker(client, task_queue="somequeue")
+  worker.register_activity(say_hello)
+  await worker.run()
+```
+
+## API Definitions
+
+A lot of loosely-defined API definitions are below. They are intentionally lacking details.
+
+High-level notes:
+
+* For demonstration/simplicity here we are leaving off a lot of proper typings (e.g. `Optional`) that the real version
+  will have
+* We leave off a lot of options on calls for this high-level proposal
+* We are not using proper function documentation, but rather inline comments for easier discussion
+* **NOTE: These will have proper docs and types when developed!**
+
+### Client
+
+Exported from `temporalio` (but probably in `temporalio/client/client.py`, exported from `temporalio.client` then
+exported from `temporalio`):
+
+```python
+class Client:
+    def __init__(
+        self,
+        addr: str,
+        *,
+        namespace: str = "default",
+        default_task_queue: str = None,
+        identity: str = None
+        # ... additional options omitted for brevity
+    ) -> None:
+        # Actually dials gRPC here
+        pass
+
+    # Raw gRPC service
+    @property
+    def service(self) -> workflowservice.WorkflowService:
+        pass
+
+    def start_workflow(
+        workflow: str,
+        *args,
+        id: str = None,
+        task_queue: str = None,
+        retry_policy: RetryPolicy = None,
+        # Note, we are embedding signal-with-start in here
+        start_signal: str = None,
+        start_signal_args=[],
```

Comment:
Not convinced we should unify these APIs

Reply:
Yup, it's an open question for discussion. I think it adds great simplicity from an API POV, but I would love to hear reasons why not besides "because we never have". I don't think the explicitness of adding `_with_signal` to the end of the name has any more value than adding `start_signal` to the end of the params.